File: HtmlMinifier.cs
// HtmlMinifier.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

class HtmlMinifier
{
    private bool RemoveComments { get; set; }
    private bool PreserveWhitespace { get; set; }
    private bool RemoveQuotes { get; set; }

    public HtmlMinifier(bool removeComments = true, bool preserveWhitespace = false, bool removeQuotes = false)
    {
        RemoveComments = removeComments;
        PreserveWhitespace = preserveWhitespace;
        RemoveQuotes = removeQuotes;
    }

    public string Minify(string html)
    {
        if (RemoveComments)
            html = Regex.Replace(html, @"<!--.*?-->", "", RegexOptions.Singleline);

        var lines = html.Split('\n');
        var sb = new System.Text.StringBuilder();
        foreach (var line in lines)
        {
            string trimmed = line.Trim();
            if (!string.IsNullOrEmpty(trimmed))
                sb.Append(trimmed);
        }
        html = sb.ToString();

        if (!PreserveWhitespace)
        {
            html = Regex.Replace(html, @">\s+<", "><");
            html = Regex.Replace(html, @" {2,}", " ");
        }

        if (RemoveQuotes)
        {
            html = Regex.Replace(html, @"=\s*""([^""]*)""", "=$1");
            html = Regex.Replace(html, @"=\s*'([^']*)'", "=$1");
        }

        html = Regex.Replace(html, @"\n\s*\n", "\n");
        return html.Trim();
    }

    static void Main(string[] args)
    {
        string inputFile = null;
        string outputFile = null;
        bool noComments = false;
        bool preserveWhitespace = false;
        bool removeQuotes = false;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "--no-comments": noComments = true; break;
                case "--preserve-whitespace": preserveWhitespace = true; break;
                case "--remove-quotes": removeQuotes = true; break;
                case "-o": outputFile = args[++i]; break;
                default: if (inputFile == null) inputFile = args[i]; break;
            }
        }

        string content;
        if (inputFile != null)
        {
            content = File.ReadAllText(inputFile);
        }
        else
        {
            using (var reader = new StreamReader(Console.OpenStandardInput()))
            {
                content = reader.ReadToEnd();
            }
        }

        var minifier = new HtmlMinifier(!noComments, preserveWhitespace, removeQuotes);
        string minified = minifier.Minify(content);

        if (outputFile != null)
            File.WriteAllText(outputFile, minified);
        else
            Console.WriteLine(minified);
    }
}

File: HtmlMinifier.java
// HtmlMinifier.java
import java.io.*;
import java.util.regex.*;

public class HtmlMinifier {
    private boolean removeComments;
    private boolean preserveWhitespace;
    private boolean removeQuotes;

    public HtmlMinifier(boolean removeComments, boolean preserveWhitespace, boolean removeQuotes) {
        this.removeComments = removeComments;
        this.preserveWhitespace = preserveWhitespace;
        this.removeQuotes = removeQuotes;
    }

    public String minify(String html) {
        if (removeComments) {
            html = html.replaceAll("<!--.*?-->", "");
        }

        String[] lines = html.split("\n");
        StringBuilder sb = new StringBuilder();
        for (String line : lines) {
            String trimmed = line.trim();
            if (!trimmed.isEmpty()) {
                sb.append(trimmed);
            }
        }
        html = sb.toString();

        if (!preserveWhitespace) {
            html = html.replaceAll(">\\s+<", "><");
            html = html.replaceAll(" {2,}", " ");
        }

        if (removeQuotes) {
            html = html.replaceAll("=\\s*\"([^\"]*)\"", "=$1");
            html = html.replaceAll("=\\s*'([^']*)'", "=$1");
        }

        html = html.replaceAll("\n\\s*\n", "\n");
        return html.trim();
    }

    public static void main(String[] args) throws IOException {
        String inputFile = null;
        String outputFile = null;
        boolean noComments = false;
        boolean preserveWhitespace = false;
        boolean removeQuotes = false;

        for (int i = 0; i < args.length; i++) {
            if ("--no-comments".equals(args[i])) noComments = true;
            else if ("--preserve-whitespace".equals(args[i])) preserveWhitespace = true;
            else if ("--remove-quotes".equals(args[i])) removeQuotes = true;
            else if ("-o".equals(args[i]) && i+1 < args.length) outputFile = args[++i];
            else if (inputFile == null) inputFile = args[i];
        }

        String content;
        if (inputFile != null) {
            try (BufferedReader reader = new BufferedReader(new FileReader(inputFile))) {
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line).append("\n");
                }
                content = sb.toString();
            }
        } else {
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line).append("\n");
                }
                content = sb.toString();
            }
        }

        HtmlMinifier minifier = new HtmlMinifier(!noComments, preserveWhitespace, removeQuotes);
        String minified = minifier.minify(content);

        if (outputFile != null) {
            try (FileWriter fw = new FileWriter(outputFile)) {
                fw.write(minified);
            }
        } else {
            System.out.println(minified);
        }
    }
}
